synapse/http/client.py

```diff
@@ -684,6 +680,18 @@ async def get_file(
         )
 
 
+def _timeout_to_request_timed_out_error(f: Failure):
+    if f.check(twisted_error.TimeoutError):
+        # the TCP connection has its own timeout (set by the 'connectTimeout' param
+        # on the Agent), which raises this exception.
+        raise RequestTimedOutError("Timeout connecting to remote server")
+    elif f.check(defer.TimeoutError):
+        # this one means that we hit our overall timeout on the request
+        raise RequestTimedOutError("Timeout waiting for response from remote server")
+
```

Comment:
I have a stack trace:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/twisted/internet/defer.py", line 1418, in _inlineCallbacks
    result = g.send(result)
  File "/usr/local/lib/python3.7/site-packages/synapse/rest/client/v1/room.py", line 260, in on_POST
    requester, event_dict, txn_id=txn_id
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/message.py", line 773, in create_and_send_nonmember_event
    ignore_shadow_ban=ignore_shadow_ban,
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/message.py", line 691, in send_nonmember_event
    requester=requester, event=event, context=context, ratelimit=ratelimit
  File "/usr/local/lib/python3.7/site-packages/synapse/util/metrics.py", line 88, in measured_func
    r = await func(self, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/synapse/handlers/message.py", line 916, in handle_new_client_event
    extra_users=extra_users,
  File "/usr/local/lib/python3.7/site-packages/synapse/replication/http/_base.py", line 196, in send_request
    result = await request_func(uri, data, headers=headers)
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 431, in post_json_get_json
    "POST", uri, headers=Headers(actual_headers), data=json_str
  File "/usr/local/lib/python3.7/site-packages/synapse/http/client.py", line 333, in request
    response = await make_deferred_yieldable(request_deferred)
twisted.web._newclient.ResponseNeverReceived: [<twisted.python.failure.Failure twisted.internet.defer.CancelledError: >]
```

which I think is a) a timeout and b) not caught here?

Reply:
"balls"

Reply:
@erikjohnston Looks like this is now handled in this PR?

Reply:
Yup